### PR TITLE
Added OAuth2 SSO login group mapping and support for nested info attributes

### LINKED ISSUE #826: Extend SSO config to be able to add users to angeltypes & set supporter rights
See also https://github.com/engelsystem/engelsystem/issues/215#issuecomment-719817208 and https://github.com/engelsystem/engelsystem/pull/772#issuecomment-727621264

### PR DESCRIPTION
**Pull request by MyIgel**
Resolves #826  (Extend SSO config to be able to add users to Angeltypes & set supporter rights).

* Allows a mapping of SSO groups to angeltypes and optional supporter rights for those
* Hides the SSO settings tab in profile if all are configured as hidden and hides the OAuth section from the profile if no OAuth is configured
* Added the option to allow registration via SSO even if registration by itself is disabled in config
* Allow info attributes to be accessed from nested arrays
  ```php
    # Info data
    [
        'ocs' => [
            'id' => 'foobar'.
            'email' => 'foo@bar.baz'.
        ]
    ]

    # Config
     'nextcloud' => [
            'name' => 'Nextcloud',
            'client_id' => '[client id]',
            'client_secret' => '[client secret',
            'url_auth' => 'https://cloud.example.com/apps/oauth2/authorize',
            'url_token' => 'https://cloud.example.com/apps/oauth2/api/v1/token',
            'url_info' => 'https://cloud.example.com/ocs/v2.php/cloud/user?format=json',
            'nested_info' => true,
            'id' => 'ocs.data.id',
            'username' => 'ocs.data.id',
            'email' => 'ocs.data.email',
        ],
  ```

### CODE REVIEW

**Comment by MyIgel** [comment]
It is also possible to set the group rights but is that a good idea? And what teams /beside heaven) would benefit from that vs. the potential security implications that allowing external systems to change permissions has on the overall usability -> @msquare, do you have an opinion on that?

**Comment by MyIgel** [comment]
Hi @weeman1337, here is another one to review 🎉

File: includes/helper/oauth_helper.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                $db->insert(
+                    'INSERT INTO UserAngelTypes (user_id, angeltype_id, confirm_user_id, supporter) VALUES (?, ?, ?, ?)',
+                    [$user->id, $team['id'], $arrived, $supporter]
+                );
+
+                continue;
+            }
+
+            if (!$supporter) {
+                continue;
```
Comment (msquare) [inline includes/helper/oauth_helper.php:72]: When this is true, it's not possible to set confirm_user_id later. Maybe do confirm_user_id first?

File: includes/helper/oauth_helper.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            $team = $ssoTeams[$groupName];
+            $userAngeltype = $currentUserAngeltypes->where('angeltype_id', $team['id'])->first();
+            $supporter = $team['supporter'];
+            $arrived = $supporter ? $user->id : null;
+
+            if (!$userAngeltype) {
+                $this->log->info(
+                    'SSO {provider}: Added to angeltype {angeltype}, arrived: {arrived}, supporter: {supporter}',
```
[the remark this answers is not in the document]
Comment (MyIgel) [reply includes/helper/oauth_helper.php:55]: Das äh stimmt :D

*dumdidum

File: includes/helper/oauth_helper.php (hunk 1 above)
Comment (MyIgel) [reply includes/helper/oauth_helper.php:72]: As you can only be confirmed automatically if you are a supporter thats imho ok that way?

Comment (msquare) [reply includes/helper/oauth_helper.php:72]: I see.

**Comment by MyIgel** [comment]
It would also be possible to use the dotted access for nested info by default and thus remove the config option -> what do you @msquare think about that?

**Review by msquare — APPROVED** [review]

**Comment by MyIgel** [comment]
Dotted access can be btw. configured using
```
     'nextcloud' => [
            'name' => 'Nextcloud',
            'client_id' => '[client id]',
            'client_secret' => '[client secret]',
            'url_auth' => 'https://cloud.example.com/apps/oauth2/authorize',
            'url_token' => 'https://cloud.example.com/apps/oauth2/api/v1/token',
            'url_info' => 'https://cloud.example.com/ocs/v2.php/cloud/user?format=json',
            'nested_info' => true,
            'id' => 'ocs.data.id',
            'username' => 'ocs.data.id',
            'email' => 'ocs.data.email',
        ],
```

**Pull request merged**